Notifications

wyattjoh opened the pull request:
## What does this PR do?

### Events

Rewrite of the events service used in Talk. Events are now the source for the subscription system which simply use the `Broker` interface to send events to the internal PubSub as well now as the instance event emitter. This allows action by action resolution to plugins implemented on the Talk nodes which can allow for some interesting features.

### Connect API

Exposes the `graph/connectors.js` via the `connect` hook.

```js
module.exports = {
  connect(connectors) {
    // use `connectors`, contents of https://github.com/coralproject/talk/blob/b758dc91cb1f1969ecd895b6059306b318995b33/graph/connectors.js#L104
  }
}
```

See https://github.com/coralproject/talk/blob/90290cfa2de88e62f687e1ed0235ba6dfe4cde26/plugins/talk-plugin-notifications/server/connect.js for an example.

### Notifications

This also introduces the notifications plugins:

- talk-plugin-notifications
- talk-plugin-notifications-category-reply

That creates a NotificationManager that creates and manages events send from the event emitter that is linked to the Graph API PubSub system.

Adding the `talk-plugin-notifications` plugin will also enable the `notifications` plugin hook. Any plugin that registers before the `talk-plugin-notifications` plugin will get picked up by.

See https://github.com/coralproject/talk/blob/8b669a31c551a042f0f079d8cfc16825673eb8f0/plugins/talk-plugin-notifications-reply/index.js for an example.

### Email Templates

Email templates can now be registered using the connect API, see https://github.com/coralproject/talk/blob/8b669a31c551a042f0f079d8cfc16825673eb8f0/plugins/talk-plugin-notifications/server/connect.js#L12-L28

wyattjoh committed:
removed events

kgardnr committed:
Update version to 4.1.0 and update LICENSE

wyattjoh committed:
blocked user actions in username change process

wyattjoh committed:
Merge branch 'master' into events

wyattjoh committed:
subscription refactor

wyattjoh committed:
Merge branch 'master' into events

wyattjoh committed:
Merge branch 'master' into events

wyattjoh committed:
Merge branch 'master' into events

wyattjoh committed:
added new plugin hook for connectors

wyattjoh committed:
refactored resolvers, cleaned

wyattjoh committed:
enhanced mailer

- added template system for dynamic re-registration of email templates
- permitted no-auth based smtp schemes (to support inbucket)
- added helper registration for dynamic components in the templates

wyattjoh committed:
initial notification draft

- implements new notification manager using connect api
- first use case of plugins using the plugin api

wyattjoh committed:
patched bug

wyattjoh committed:
Merge branch 'master' into events

wyattjoh committed:
added docs, added new logger

wyattjoh committed:
added mutations

wyattjoh committed:
patched test

wyattjoh committed:
Merge branch 'master' into events

alinhle commented on graph/resolvers/comment.js:
```diff
@@ -29,15 +31,8 @@ const Comment = {
-      return null;
-    }
-
+  replyCount: property('reply_count'),
```
Interesting interesting

wyattjoh committed:
emails loaded from graph

- introduced graphql method on context
- mailer will try to load the email from the graph

alinhle commented on serve.js:
```diff
@@ -79,6 +80,9 @@ async function onListening() {
  * Start the app.
  */
 async function serve({ jobs = false, websockets = false } = {}) {
+  // Run the deffered plugins.
```
Typo :-P

alinhle committed:
Add changes

wyattjoh committed:
fixed runtime errors

wyattjoh committed:
replaced json -> hjson loader

wyattjoh committed:
lint

wyattjoh committed:
Merge branch 'master' into events

wyattjoh committed:
fixed test

wyattjoh committed:
fixes based on feedback

wyattjoh committed:
Merge branch 'master' into events

wyattjoh committed:
fixed import cycle

alinhle committed:
Merge branch 'refactor-profile' into events

alinhle committed:
Merge branch 'refactor-profile' into events

alinhle committed:
WIP

wyattjoh committed:
Update router.js

wyattjoh committed:
Merge branch 'master' into events

wyattjoh committed:
spelling

alinhle committed:
Implemented notification frontend functionality

alinhle committed:
Merge branch 'events' of ssh://github.com/coralproject/talk into events

wyattjoh committed:
mail cleanup

wyattjoh committed:
improved mailer

wyattjoh committed:
removed console.log

alinhle committed:
Rename fragment

alinhle committed:
Merge branch 'events' of ssh://github.com/coralproject/talk into events

alinhle committed:
Merge branch 'refactor-profile' into events

alinhle committed:
Merge branch 'refactor-profile' into events

alinhle committed:
Fix import path

alinhle committed:
Implement styling

alinhle committed:
Refactor toggle

wyattjoh committed:
replaced localStorage state events with postMessage

wyattjoh committed:
linting

wyattjoh committed:
added sessionStorage fallback

wyattjoh committed:
Merge branch 'master' into events

wyattjoh committed:
merged lockfiles

wyattjoh committed:
fixed permission bug

wyattjoh committed:
cleaned up notifications manager

wyattjoh committed:
added support for the singleSlot

wyattjoh committed:
added singleSlot on the commentContent

wyattjoh committed:
replaced singleSlot with slotSize

wyattjoh committed:
Merge branch 'master' into events

wyattjoh committed:
implemented unsubscribe button

wyattjoh committed:
Update Slot.js

wyattjoh committed:
Update plugins.js

wyattjoh committed:
linting

alinhle committed:
Remove redudant request for user object

alinhle committed:
Refactor postMessage service

alinhle committed:
Remove deprecated part <3

alinhle committed:
Bye bye old code

alinhle committed:
Fix closedAt wrong datatype

alinhle committed:
fix jsdoc

alinhle committed:
fix jsdoc 2

alinhle committed:
Purge config.auth_token from redux on logout

wyattjoh committed:
Merge branch 'master' into events

wyattjoh committed:
fixed linting error
